reconciler/healthchecks/healthchecks.go

```diff
@@ -0,0 +1,295 @@
+	}
+	c.infos[id] = info
+	return nil
+}
+
+func (c *checker) updateHealthCheck(info *checkInfo, check *types.VirtualService_HealthCheck) {
+	info.healthCheck = check
+
+	// stop all existing health check goroutines
+	for server, stopCh := range info.serverStopChs {
+		close(stopCh)
+		delete(info.serverStopChs, server)
+	}
+
+	if check == nil {
+		return
+	}
+
+	// kick off new health checks with the updated check
+	for server, status := range info.serverStatuses {
+		stopCh := make(chan struct{})
+		info.serverStopChs[server] = stopCh
+		go status.healthCheck(stopCh, server, check)
+	}
+}
+
+func (c *checker) AddServer(id, server string) {
+	c.Lock()
+	defer c.Unlock()
+
+	info, ok := c.infos[id]
+	if !ok {
+		// initialise health check if not set yet
+		c.Unlock()
+		if err := c.SetHealthCheck(id, nil); err != nil {
+			panic(err)
+		}
+		c.Lock()
+		info = c.infos[id]
+	}
+	if _, ok := info.serverStatuses[server]; ok {
+		// server already added
+		return
+	}
+
+	status := &checkStatus{
+		state: serverDown,
+		count: 0,
+	}
+	info.serverStatuses[server] = status
+
+	if info.healthCheck == nil {
+		//  nothing more to do, return
+		return
+	}
+
+	// kick off health check
+	stopCh := make(chan struct{})
+	info.serverStopChs[server] = stopCh
+	go status.healthCheck(stopCh, server, info.healthCheck)
+}
+
+func (c *checker) RemServer(id, server string) {
+	c.Lock()
+	defer c.Unlock()
+
+	info, ok := c.infos[id]
+	if !ok {
+		// nothing to remove
+		return
+	}
+	if _, ok := info.serverStatuses[server]; !ok {
+		// server already removed
+		return
+	}
+
+	if stopCh, ok := info.serverStopChs[server]; ok {
+		close(stopCh)
+	}
+	delete(info.serverStopChs, server)
+	delete(info.serverStatuses, server)
+}
+
+func (c *checker) Stop() {
+	c.Lock()
+	defer c.Unlock()
+
+	for _, info := range c.infos {
+		c.updateHealthCheck(info, nil)
+	}
+}
+
+// healthCheck the servers and update their status
+func (s *checkStatus) healthCheck(stopCh <-chan struct{}, server string, healthCheck *types.VirtualService_HealthCheck) {
+	per, _ := ptypes.Duration(healthCheck.Period)
+	t := time.NewTicker(per)
+	defer t.Stop()
+	for {
+		select {
+		case <-t.C:
+			s.performHealthCheck(server, healthCheck)
+		case <-stopCh:
+			return
+		}
+	}
+}
+
+func (s *checkStatus) performHealthCheck(server string, healthCheck *types.VirtualService_HealthCheck) {
+	checkURL, err := url.Parse(healthCheck.Endpoint)
+	if err != nil {
+		panic(err)
+	}
+	if checkURL.Scheme != "http" {
+		panic("unsupported health check scheme " + checkURL.Scheme)
+	}
+
+	// Create a custom transport so we don't reuse prior connections, which might hide connectivity problems.
+	tr := &http.Transport{
+		DisableKeepAlives: true,
+	}
+	timeout, err := ptypes.Duration(healthCheck.Timeout)
+	if err != nil {
+		panic(err)
+	}
+	client := http.Client{
+		Transport: tr,
+		Timeout:   timeout,
+	}
+
+	serverURL, err := url.Parse(fmt.Sprintf("http://%s:%s%s", server, checkURL.Port(), checkURL.Path))
```

Comment:
Is there a nicer way to do this using the existing checkURL?

Reply:
It's a bit ugly mutating the previous url struct fields, so I've left this as is.